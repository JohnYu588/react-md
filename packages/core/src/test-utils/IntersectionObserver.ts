export class IntersectionObserverMock implements IntersectionObserver {
  root: Document | Element | null;
  rootMargin: string;
  thresholds: readonly number[];

  elements: Set<Element>;

  constructor(
    public callback: IntersectionObserverCallback,
    options: IntersectionObserverInit = {}
  ) {
    this.root = options.root || null;
    this.rootMargin = options.rootMargin || "";
    this.thresholds =
      typeof options.threshold === "number"
        ? [options.threshold]
        : options.threshold ?? [];

    this.elements = new Set();
  }

  observe(target: Element): void {
    this.elements.add(target);

    this.callback(this.takeRecords(), this);
  }

  unobserve(target: Element): void {
    this.elements.delete(target);
  }

  takeRecords(): IntersectionObserverEntry[] {
    return [...this.elements].map<IntersectionObserverEntry>((target) => ({
      time: Date.now(),
      target,
      boundingClientRect: target.getBoundingClientRect(),
      intersectionRatio: 0,
      intersectionRect: target.getBoundingClientRect(),
      isIntersecting: false,
      rootBounds:
        this.root && "getBoundingClientRect" in this.root
          ? this.root.getBoundingClientRect()
          : null,
    }));
  }

  disconnect(): void {
    this.elements.clear();
  }
}
