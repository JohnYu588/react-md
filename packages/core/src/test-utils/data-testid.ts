/// <reference types="react" />

declare module "react" {
  interface HTMLAttributes<T>
    extends React.AriaAttributes,
      React.DOMAttributes<T> {
    "data-testid"?: string | number;
  }
}
