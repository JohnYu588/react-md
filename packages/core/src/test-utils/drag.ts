import { fireEvent } from "@testing-library/dom";
import type { MouseEvent } from "react";
import { wait } from "../utils/wait.js";

interface XYCoords {
  x: number;
  y: number;
}

interface BaseDragOptions {
  to?: XYCoords | Element;
  delta?: XYCoords;
  steps?: number;
  duration?: number;
}

type DragOptions = BaseDragOptions &
  ({ to: XYCoords | Element; delta?: never } | { delta: XYCoords; to?: never });

const getElementClientCenter = (element: Element): XYCoords => {
  const { left, top, width, height } = element.getBoundingClientRect();
  return {
    x: left + width / 2,
    y: top + height / 2,
  };
};

const getCoords = (elementOrCoords: Element | XYCoords): XYCoords =>
  "x" in elementOrCoords && "y" in elementOrCoords
    ? elementOrCoords
    : getElementClientCenter(elementOrCoords);

/**
 * This is internal for now. No idea if it's actually worthwhile though since
 * you have to mock the `getBoundingClientRect()` to work.
 *
 * @internal
 * @see https://github.com/testing-library/testing-library-docs/blob/c20f8a8931e9f0a158fda9f01bbea8cb45ed1651/docs/example-drag.mdx
 */
export async function drag(
  element: Element,
  options: DragOptions
): Promise<void> {
  const { delta, to: inTo, steps = 20, duration = 500 } = options;
  const from = getElementClientCenter(element);
  const to = delta
    ? {
        x: from.x + delta.x,
        y: from.y + delta.y,
      }
    : getCoords(inTo);

  const step: XYCoords = {
    x: (to.x - from.x) / steps,
    y: (to.y - from.y) / steps,
  };

  const currentEvent = {
    clientX: from.x,
    clientY: from.y,
    button: 0,
    altKey: false,
    ctrlKey: false,
    metaKey: false,
    shiftKey: false,
  } satisfies Partial<MouseEvent>;

  fireEvent.mouseEnter(element, currentEvent);
  fireEvent.mouseOver(element, currentEvent);
  fireEvent.mouseMove(element, currentEvent);
  fireEvent.mouseDown(element, currentEvent);
  fireEvent.dragStart(element, currentEvent);

  const delay = duration / steps;
  for (let i = 0; i < steps; i++) {
    currentEvent.clientX += step.x;
    currentEvent.clientY += step.y;
    await wait(delay);
    fireEvent.mouseMove(element, currentEvent);
  }

  fireEvent.mouseUp(element, currentEvent);
  fireEvent.dragEnd(element, currentEvent);
}
