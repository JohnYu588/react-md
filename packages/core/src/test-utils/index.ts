export * from "@testing-library/react";
export * from "@testing-library/user-event";
export * from "./mocks/match-media.js";
export * from "./render.js";
export * from "./utils/resize-observer.js";
