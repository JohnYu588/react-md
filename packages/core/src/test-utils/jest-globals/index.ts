export * from "./match-media.js";
export * from "./resize-observer.js";
export * from "./timers.js";
