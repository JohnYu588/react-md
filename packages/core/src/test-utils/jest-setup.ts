import { beforeEach } from "@jest/globals";
import { INTERACTION_CONFIG } from "../interaction/config.js";

beforeEach(() => {
  // set the mode to `none` in tests since ripples require
  // `getBoundingClientRect()` to create correct CSS. You'll either see warnings
  // in the console around invalid css values or `NaN`.
  INTERACTION_CONFIG.mode = "none";
});
