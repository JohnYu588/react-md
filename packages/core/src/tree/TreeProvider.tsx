"use client";
import type { ReactElement, ReactNode } from "react";
import { createContext, useContext, useMemo } from "react";
import type { NonNullRef } from "../types.js";
import type { DefaultTreeItemNode, TreeData, TreeItemNode } from "./types.js";
import type { TreeExpansion } from "./useTreeExpansion.js";
import type { TreeSelection } from "./useTreeSelection.js";

/**
 * When this is set to `"auto"`, clicking on a tree item with a mouse or with
 * the Enter/Space keys will select and expand/collapse the tree item. The user
 * can still use the ArrowLeft and ArrowRight keys to expand/collapse nested
 * tree items.
 *
 * When this is set to `"manual"`, clicking on a tree item with a mouse or with
 * the Enter/Space keys will only select that tree item. The user must use the
 * ArrowLeft and ArrowRight keys to expand/collapse nested tree items.
 *
 * The main use-case for the `"manual"` setting is creating a navigation tree
 * that requires the user to click on an icon or a button to expand the child
 * items.
 *
 * @remarks \@since 6.0.0
 */
export type TreeExpansionMode = "auto" | "manual";

/**
 * @internal
 * @remarks \@since 6.0.0
 */
export interface TreeItemMetadataLookup {
  /**
   * This is used for the keyboard movement behavior for trees that allow the
   * `ArrowRight` key for tree items that have children.
   */
  expandable: Record<string, boolean>;

  /**
   * This is a quick lookup if a tree item is disabled since it's possible the
   * custom `renderer` sets the `disabled` state and not the `TreeData`.
   */
  disabledItems: Record<string, boolean>;

  /**
   * A lookup of element `id` (DOM id) for each tree item to the `itemId`. This
   * is a quick way to figure out which item is being interacted with from a
   * keyboard event without needed to add `data-*` attributes.
   */
  elementToItem: Record<string, string>;

  /**
   * A lookup of tree item `itemId` to the element `id` (DOM id). This is used
   * to be able to find a parent tree item when the `ArrowLeft` key is pressed
   * to focus that parent item.
   */
  itemToElement: Record<string, string>;
}

/**
 * The tree context is mostly a convenience API for implementing a custom tree
 * item that requires the {@link TreeExpansion}, {@link TreeSelection} and
 * {@link TreeData} to work.
 *
 * The other properties on the context are most likely for internal use only.
 *
 * @remarks \@since 6.0.0
 */
export interface TreeContext<T extends TreeItemNode = DefaultTreeItemNode>
  extends TreeExpansion,
    TreeSelection {
  data: TreeData<T>;
  rootId: string | null;
  disableTransition: boolean;
  temporaryChildItems: boolean;

  /** @internal */
  expanderLeft: boolean;
  /** @internal */
  expanderIcon: ReactNode;
  /** @internal */
  expansionMode: TreeExpansionMode;
  /** @internal */
  metadataLookup: NonNullRef<TreeItemMetadataLookup>;
}

// Allow the hook to correct typecast this instead
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const context = createContext<TreeContext<any> | undefined>(undefined);
context.displayName = "Tree";
const { Provider } = context;

/**
 * This can be used for a custom tree item renderer.
 *
 * @see The `Tree` component for an example.
 * @remarks \@since 6.0.0
 */
export function useTreeContext<
  T extends TreeItemNode = DefaultTreeItemNode,
>(): TreeContext<T> {
  const value = useContext(context);
  if (!value) {
    throw new Error("Cannot find a parent Tree component");
  }

  return value;
}

/**
 * @internal
 * @remarks \@since 6.0.0
 */
export interface TreeProviderProps<T extends TreeItemNode = DefaultTreeItemNode>
  extends TreeContext<T> {
  children: ReactNode;
}

/**
 * **Client Component**
 *
 * @internal
 * @remarks \@since 6.0.0
 */
export function TreeProvider<T extends TreeItemNode = DefaultTreeItemNode>(
  props: TreeProviderProps<T>
): ReactElement {
  const {
    children,
    data,
    rootId,
    multiSelect,
    expanderIcon,
    expanderLeft,
    expansionMode,
    metadataLookup,
    disableTransition,
    temporaryChildItems,
    expandedIds,
    selectedIds,
    expandMultipleTreeItems,
    toggleTreeItemExpansion,
    selectMultipleTreeItems,
    toggleTreeItemSelection,
  } = props;
  const value = useMemo<TreeContext<T>>(
    () => ({
      data,
      rootId,
      multiSelect,
      expanderIcon,
      expanderLeft,
      expansionMode,
      metadataLookup,
      disableTransition,
      temporaryChildItems,
      expandedIds,
      selectedIds,
      expandMultipleTreeItems,
      toggleTreeItemExpansion,
      selectMultipleTreeItems,
      toggleTreeItemSelection,
    }),
    [
      data,
      disableTransition,
      expandMultipleTreeItems,
      expandedIds,
      expanderIcon,
      expanderLeft,
      expansionMode,
      metadataLookup,
      multiSelect,
      rootId,
      selectMultipleTreeItems,
      selectedIds,
      temporaryChildItems,
      toggleTreeItemExpansion,
      toggleTreeItemSelection,
    ]
  );

  return <Provider value={value}>{children}</Provider>;
}
