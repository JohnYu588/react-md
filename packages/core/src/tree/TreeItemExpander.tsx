import { cnb } from "cnbuilder";
import type { MouseEventHandler, ReactElement, ReactNode } from "react";
import { cloneElement, isValidElement } from "react";
import { IconRotator, useIcon } from "../icon";
import { useTreeContext } from "./TreeProvider";

/**
 * @internal
 * @remarks \@since 6.0.0 Updated to use the latest API
 */
export interface TreeItemExpanderProps {
  /** @defaultValue `false` */
  isLeft?: boolean;
  itemId: string;
  addon: ReactNode;
  disabled: boolean;
  expanded: boolean;
  isLeafNode: boolean;
  className?: string;
}

/**
 * @internal
 * @remarks \@since 6.0.0 Updated to support the new `expansionMode` behavior.
 */
export function TreeItemExpander(props: TreeItemExpanderProps): ReactElement {
  const {
    itemId,
    isLeft = false,
    addon,
    expanded,
    disabled,
    isLeafNode,
    className,
  } = props;
  const {
    expanderIcon,
    expanderLeft,
    expansionMode,
    disableTransition,
    toggleTreeItemExpansion,
  } = useTreeContext();

  const icon = useIcon("dropdown", expanderIcon);
  if (isLeafNode || expanderLeft !== isLeft) {
    if (isValidElement<{ className?: string }>(addon)) {
      return cloneElement(addon, { className });
    }

    return <>{addon}</>;
  }

  const isCloneable = isValidElement(icon);
  let clickProps: { onClick: MouseEventHandler<Element> } | undefined;
  if (expansionMode === "manual" && !disabled) {
    clickProps = {
      onClick(event) {
        // do not select the tree item if the icon is clicked
        event.stopPropagation();
        event.preventDefault();

        toggleTreeItemExpansion(itemId);
      },
    };
  }

  return (
    <>
      <IconRotator
        className={cnb(!addon && className)}
        {...(!isCloneable ? clickProps : undefined)}
        rotated={expanded}
        forceIconWrap={!isCloneable}
        disableTransition={disableTransition}
      >
        {clickProps && isCloneable ? cloneElement(icon, clickProps) : icon}
      </IconRotator>
      {addon}
    </>
  );
}
