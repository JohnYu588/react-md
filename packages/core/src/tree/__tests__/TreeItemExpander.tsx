import { fireEvent, render } from "../../test-utils";

import { TreeItemExpander } from "../TreeItemExpander";
import { TreeProvider } from "../TreeProvider";

describe("TreeItemExpander", () => {
  it("should throw an error if not wrapped in a TreeProvider", () => {
    const error = jest.spyOn(console, "error").mockImplementation(() => {});
    expect(() =>
      render(
        <TreeItemExpander
          itemId="item-1-id"
          addon={null}
          disabled={false}
          expanded={false}
          isLeafNode={false}
        />
      )
    ).toThrow("Cannot find a parent Tree component");

    error.mockRestore();
  });

  it("should force wrapping the icon in a span and move the click props to the span if the icon is not cloneable", () => {
    const toggleTreeItemExpansion = jest.fn();
    const { container } = render(
      <TreeProvider
        temporaryChildItems={false}
        data={{}}
        disableTransition={false}
        expandedIds={new Set()}
        expanderIcon="icon"
        expanderLeft={false}
        expandMultipleTreeItems={jest.fn()}
        expansionMode="manual"
        metadataLookup={{
          current: {
            disabledItems: {},
            elementToItem: {},
            expandable: {},
            itemToElement: {},
          },
        }}
        rootId={null}
        selectedIds={new Set()}
        selectMultipleTreeItems={jest.fn()}
        toggleTreeItemExpansion={toggleTreeItemExpansion}
        toggleTreeItemSelection={jest.fn()}
        multiSelect={false}
      >
        <TreeItemExpander
          data-testid="icon"
          itemId="item-1-id"
          addon={null}
          disabled={false}
          expanded={false}
          isLeafNode={false}
        />
      </TreeProvider>
    );

    const iconWrapper = container.firstElementChild;
    if (!iconWrapper) {
      throw new Error();
    }
    expect(iconWrapper).toBeInstanceOf(HTMLSpanElement);
    expect(container).toMatchSnapshot();

    fireEvent.click(iconWrapper);
    expect(toggleTreeItemExpansion).toHaveBeenCalledWith("item-1-id");
  });
});
