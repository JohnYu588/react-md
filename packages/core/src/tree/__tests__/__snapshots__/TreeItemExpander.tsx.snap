// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeItemExpander should force wrapping the icon in a span and move the click props to the span if the icon is not cloneable 1`] = `
<div>
  <span
    class="rmd-icon-rotator rmd-icon-rotator--animate"
  >
    icon
  </span>
</div>
`;
